### DISCUSSION TITLE
Issues related to loading custom lemmas via lookups

### DISCUSSION THREAD

**Discussion by ghost** [explosion/spaCy#11259] [Help: Other Questions] [answered]
Hi,
I had written a function to add custom lemmas/pos tags/vocabulary that used to work in the previous version but doesn't work anymore. I've tried various solutions provided online but they don't work either. 
What my code should do is, add a token and its custom lemma for words that are not correctly lemmatized, load that into the pipeline and then get the correct lemma when called for it.

The piece of code that does this (referenced from https://stackoverflow.com/questions/63260356/add-rules-to-spacy-lemmatization)
```import spacy
from spacy.lookups import load_lookups

nlp = spacy.load("de_core_news_lg")

assert nlp("gällt")[0].lemma_ in ("gällt", "gällen")
# print(nlp("gällt")[0].pos_) # VERB
# print(nlp("gällt")[0].pos) # 100

# 1. Exception for rule-based lemmatizer (with tagger)

# reload to start with a clean lemma cache
nlp = spacy.load("de_core_news_lg", disable=["parser", "ner"])
config = {"mode": "rule", "overwrite": True}
nlp.remove_pipe('lemmatizer')
nlp.add_pipe("lemmatizer", config=config)

# add an exception for "ups" as with POS NOUN or VERB
nlp.get_pipe('lemmatizer').lookups.get_table("lemma_exc")["noun"]["gällt"] = ["gällen"]
nlp.get_pipe('lemmatizer').lookups.get_table("lemma_exc")["verb"]["gällt"] = ["gällen"]
print(nlp("gällt")[0].lemma_)  # gällt
assert nlp("gällt")[0].lemma_ == "gällen"  # fails

# 2. New entry for lookup lemmatizer (without tagger)

nlp = spacy.load("de_core_news_lg", disable=["tagger", "parser", "ner"])
nlp.vocab.lookups.get_table("lexeme_norm")["gällt"] = "gällen"
assert nlp("gällt")[0].lemma_ == "gällen"
```

Error messages that keep popping up:

```
Traceback (most recent call last):
  File "/Users/dev/PycharmProjects/python/scripts/spacy_lemmatization.py", line 19, in <module>
    nlp.get_pipe('lemmatizer').lookups.get_table("lemma_exc")["noun"]["gällt"] = ["gällen"]
  File "/Users/dev/.pyenv/versions/3.8.10/envs/venv381/lib/python3.8/site-packages/spacy/lookups.py", line 232, in get_table
    raise KeyError(Errors.E159.format(name=name, tables=self.tables))
KeyError: "[E159] Can't find table 'lemma_exc' in lookups. Available tables: []"
```

We also tried using other lookup tables such as 'lexeme_norm' but they all give errors.
Another error we came across was:
```attributeerror: 'edittreelemmatizer' object has no attribute 'lookups' spacy```

Pleasee let me know how this can be solved. Thank you!

**Comment by adrianeboyd** [comment, marked as answer]
If it's the standard non-trainable lemmatizer, the `lemma_*` tables are in the lemmatizer component under `nlp.get_pipe("lemmatizer").lookups`. The `lexeme_norm` table is still in the vocab under `nlp.vocab.lookups`.

Here are some notes on loading different types of lemmatizers in v3 pipelines (scroll down to "Switch from trainable lemmatizer to default lemmatizer"): https://spacy.io/models#design-modify

Ah, I think I see what's going on. There isn't a rule-based lemmatizer for German, and you're missing the initialization step that would show errors when it can't find the tables to load from `spacy-lookups-data`.

```python
lemmatizer = nlp.add_pipe("lemmatizer", config=config)
lemmatizer.initialize()
```

**Reply by ghost** [reply]
Thank you!
